pkg/controlplane/controller/apiserverleasegc/gc_controller.go

```diff
@@ -0,0 +1,136 @@
+	"k8s.io/client-go/tools/cache"
+
+	"k8s.io/klog/v2"
+)
+
+// Controller deletes expired apiserver leases.
+type Controller struct {
+	kubeclientset kubernetes.Interface
+
+	leaseLister   listers.LeaseLister
+	leaseInformer cache.SharedIndexInformer
+	leasesSynced  cache.InformerSynced
+
+	leaseNamespace string
+
+	gcCheckPeriod time.Duration
+}
+
+// NewAPIServerLeaseGC creates a new Controller.
+func NewAPIServerLeaseGC(clientset kubernetes.Interface, gcCheckPeriod time.Duration, leaseNamespace, leaseLabelSelector string) *Controller {
+	// we construct our own informer because we need such a small subset of the information available.
+	// Just one namespace with label selection.
+	leaseInformer := informers.NewFilteredLeaseInformer(
+		clientset,
+		leaseNamespace,
+		0,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(listOptions *metav1.ListOptions) {
+			listOptions.LabelSelector = leaseLabelSelector
+		})
+	return &Controller{
+		kubeclientset:  clientset,
+		leaseLister:    listers.NewLeaseLister(leaseInformer.GetIndexer()),
+		leaseInformer:  leaseInformer,
+		leasesSynced:   leaseInformer.HasSynced,
+		leaseNamespace: leaseNamespace,
+		gcCheckPeriod:  gcCheckPeriod,
+	}
+}
+
+// Run starts one worker.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer klog.Infof("Shutting down apiserver lease garbage collector")
+
+	klog.Infof("Starting apiserver lease garbage collector")
+
+	// we have a personal informer that is narrowly scoped, start it.
+	go c.leaseInformer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.leasesSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.gc, c.gcCheckPeriod, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) gc() {
+	leases, err := c.leaseLister.Leases(c.leaseNamespace).List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Error while listing apiserver leases: %v", err)
+		return
+	}
+	for _, lease := range leases {
+		// evaluate lease from cache
+		if !isLeaseExpired(lease) {
+			continue
+		}
+		// double check latest lease from apiserver before deleting
+		lease, err := c.kubeclientset.CoordinationV1().Leases(c.leaseNamespace).Get(context.TODO(), lease.Name, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("Error getting lease: %v", err)
+			continue
+		}
+		if errors.IsNotFound(err) || lease == nil {
+			// the lease was deleted by the same GC controller in another apiserver
+			continue
+		}
+		// evaluate lease from apiserver
+		if !isLeaseExpired(lease) {
+			continue
+		}
+		if err := c.kubeclientset.CoordinationV1().Leases(c.leaseNamespace).Delete(context.TODO(), lease.Name, metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+			// If we get a 404, the lease was deleted by the same GC controller
+			// in another apiserver
+			if !errors.IsNotFound(err) {
```

Comment:
It's probably not needed to call IsNotFound twice :)

Reply:
now I see why @caesarxuchao asks me to break the long lines.. this is what happens :)